### ISSUE TITLE
OPNsense 19.7.10-amd64 openvpn server seem to be broken

### ISSUE THREAD

**Issue by user111222333** [opnsense/core#3899]
Hello Dev Team, 
It appears after today's update, it is not possible to utilize the openvpn service running on OPNsense.

The most basic vanila/default setup result in client(s) not able to connect. 
Logs show that openvpn demon gets the first packet from the client  when negotiating the TLS session, but server never responds back and client resets the connection and tries again. 

There is additional issues with the openvpn set-up wizard, when wizard is used, and script auto adds the firewall rules, instruction for a protocol configures ipv4/ipv6 while global setting for the OPNSense is set to prefer ipv4 only. A user has to manual edit the rule and with out changing any settings, just simply click 'save' button at the bottom of the rule edit page and the ip protocol version magically changes itself to ipv4 all by itself. 

Please not, firewall rules are set, OpenVPN rule for outgoing connections set to any:any.
Logs show that incoming connections are reaching to openvpn service... and it still does not work any longer. 

  

Here's the details: 

Versions: 	
OPNsense 19.7.10-amd64
FreeBSD 11.2-RELEASE-p16-HBSD
OpenSSL 1.0.2u 20 Dec 2019 

Logs on a server: 

Jan 29 20:28:15 	openvpn[54295]: MANAGEMENT: Client disconnected
Jan 29 20:28:15 	openvpn[54295]: MANAGEMENT: CMD 'quit'
Jan 29 20:28:15 	openvpn[54295]: MANAGEMENT: CMD 'status 2'
Jan 29 20:28:15 	openvpn[54295]: MANAGEMENT: Client connected from /var/etc/openvpn/server1.sock
Jan 29 20:27:13 	openvpn[54295]: MANAGEMENT: Client disconnected
Jan 29 20:27:13 	openvpn[54295]: MANAGEMENT: CMD 'quit'
Jan 29 20:27:13 	openvpn[54295]: MANAGEMENT: CMD 'status 2'
Jan 29 20:27:13 	openvpn[54295]: MANAGEMENT: Client connected from /var/etc/openvpn/server1.sock
Jan 29 20:26:11 	openvpn[54295]: MANAGEMENT: Client disconnected
Jan 29 20:26:11 	openvpn[54295]: MANAGEMENT: CMD 'quit'
Jan 29 20:26:11 	openvpn[54295]: MANAGEMENT: CMD 'status 2'
Jan 29 20:26:11 	openvpn[54295]: MANAGEMENT: Client connected from /var/etc/openvpn/server1.sock
Jan 29 20:25:09 	openvpn[54295]: MANAGEMENT: Client disconnected

Logs on a client side: 

Wed Jan 29 19:54:05 2020 TCP/UDP: Preserving recently used remote address: [AF_INET]10.0.90.102:1194
Wed Jan 29 19:54:05 2020 UDP link local (bound): [AF_INET][undef]:0
Wed Jan 29 19:54:05 2020 UDP link remote: [AF_INET]10.0.90.102:1194
Wed Jan 29 19:55:05 2020 TLS Error: TLS key negotiation failed to occur within 60 seconds (check your network connectivity)
Wed Jan 29 19:55:05 2020 TLS Error: TLS handshake failed
Wed Jan 29 19:55:05 2020 SIGUSR1[soft,tls-error] received, process restarting
Wed Jan 29 19:55:10 2020 TCP/UDP: Preserving recently used remote address: [AF_INET]10.0.90.102:1194
Wed Jan 29 19:55:10 2020 UDP link local (bound): [AF_INET][undef]:0
Wed Jan 29 19:55:10 2020 UDP link remote: [AF_INET]10.0.90.102:1194
Wed Jan 29 19:56:10 2020 TLS Error: TLS key negotiation failed to occur within 60 seconds (check your network connectivity)
Wed Jan 29 19:56:10 2020 TLS Error: TLS handshake failed
Wed Jan 29 19:56:10 2020 SIGUSR1[soft,tls-error] received, process restarting
Wed Jan 29 19:56:15 2020 TCP/UDP: Preserving recently used remote address: [AF_INET]10.0.90.102:1194
Wed Jan 29 19:56:15 2020 UDP link local (bound): [AF_INET][undef]:0
Wed Jan 29 19:56:15 2020 UDP link remote: [AF_INET]10.0.90.102:1194
Wed Jan 29 19:57:15 2020 TLS Error: TLS key negotiation failed to occur within 60 seconds (check your network connectivity)
Wed Jan 29 19:57:15 2020 TLS Error: TLS handshake failed

Auto-generated config for clients missing IOS config, android config, unified .ovpn config... 
Published docs for setting up openvpn instance seem outdated/irrelevant to the current version of OPNsense... 

Please advise what a best course of action for a user to set up a working openvpn instance running on OPNsense. 

Thank you.

**Comment by mimugmail** [comment]
So it worked with 19.7.9? From the logs it seems like blocked connection

**Comment by user111222333** [comment]
It does seem like a blocked connection, but all the firewall rules are in place... 

Here more verbose log output....


Jan 29 21:00:23 	openvpn[88704]: MANAGEMENT: Client disconnected
Jan 29 21:00:23 	openvpn[88704]: MANAGEMENT: CMD 'quit'
Jan 29 21:00:23 	openvpn[88704]: MANAGEMENT: CMD 'status 2'
Jan 29 21:00:22 	openvpn[88704]: MANAGEMENT: Client connected from /var/etc/openvpn/server1.sock
Jan 29 20:59:21 	openvpn[88704]: MANAGEMENT: Client disconnected
Jan 29 20:59:21 	openvpn[88704]: MANAGEMENT: CMD 'quit'
Jan 29 20:59:20 	openvpn[88704]: MANAGEMENT: CMD 'status 2'
Jan 29 20:59:20 	openvpn[88704]: MANAGEMENT: Client connected from /var/etc/openvpn/server1.sock
Jan 29 20:58:26 	openvpn[88704]: Initialization Sequence Completed
Jan 29 20:58:26 	openvpn[88704]: IFCONFIG POOL: base=10.0.8.2 size=252, ipv6=0
Jan 29 20:58:26 	openvpn[88704]: MULTI: multi_init called, r=256 v=256
Jan 29 20:58:26 	openvpn[88704]: UDPv4 link remote: [AF_UNSPEC]
Jan 29 20:58:26 	openvpn[88704]: UDPv4 link local (bound): [AF_INET]10.0.90.102:1194
Jan 29 20:58:26 	openvpn[88704]: Socket Buffers: R=[42080->42080] S=[57344->57344]
Jan 29 20:58:26 	openvpn[88704]: Could not determine IPv4/IPv6 protocol. Using AF_INET
Jan 29 20:58:26 	openvpn[88704]: Data Channel MTU parms [ L:1622 D:1450 EF:122 EB:406 ET:0 EL:3 ]
Jan 29 20:58:26 	openvpn[88704]: /usr/local/etc/inc/plugins.inc.d/openvpn/ovpn-linkup ovpns1 1500 1622 10.0.8.1 255.255.255.0 init
Jan 29 20:58:26 	openvpn[88704]: /sbin/route add -net 10.0.8.0 10.0.8.2 255.255.255.0
Jan 29 20:58:26 	openvpn[88704]: /sbin/ifconfig ovpns1 10.0.8.1 10.0.8.2 mtu 1500 netmask 255.255.255.0 up
Jan 29 20:58:26 	openvpn[88704]: do_ifconfig, tt->did_ifconfig_ipv6_setup=0
Jan 29 20:58:26 	openvpn[88704]: TUN/TAP device /dev/tun1 opened
Jan 29 20:58:26 	openvpn[88704]: TUN/TAP device ovpns1 exists previously, keep at program end
Jan 29 20:58:26 	openvpn[88704]: TLS-Auth MTU parms [ L:1622 D:1172 EF:78 EB:0 ET:0 EL:3 ]
Jan 29 20:58:26 	openvpn[88704]: Incoming Control Channel Authentication: Using 256 bit message hash 'SHA256' for HMAC authentication
Jan 29 20:58:26 	openvpn[88704]: Outgoing Control Channel Authentication: Using 256 bit message hash 'SHA256' for HMAC authentication
Jan 29 20:58:26 	openvpn[88704]: Diffie-Hellman initialized with 2048 bit key
Jan 29 20:58:26 	openvpn[88704]: NOTE: the current --script-security setting may allow this configuration to call user-defined scripts
Jan 29 20:58:26 	openvpn[88704]: MANAGEMENT: unix domain socket listening on /var/etc/openvpn/server1.sock
Jan 29 20:58:26 	openvpn[76801]: library versions: OpenSSL 1.0.2u 20 Dec 2019, LZO 2.10
Jan 29 20:58:26 	openvpn[76801]: auth_user_pass_file = '[UNDEF]'
Jan 29 20:58:26 	openvpn[76801]: pull = DISABLED
Jan 29 20:58:26 	openvpn[76801]: client = DISABLED
Jan 29 20:58:26 	openvpn[76801]: port_share_port = '[UNDEF]'
Jan 29 20:58:26 	openvpn[76801]: port_share_host = '[UNDEF]'
Jan 29 20:58:26 	openvpn[76801]: auth_token_lifetime = 0
Jan 29 20:58:26 	openvpn[76801]: auth_token_generate = DISABLED
Jan 29 20:58:26 	openvpn[76801]: auth_user_pass_verify_script_via_file = DISABLED
Jan 29 20:58:26 	openvpn[76801]: max_routes_per_client = 256
Jan 29 20:58:26 	openvpn[76801]: max_clients = 2
Jan 29 20:58:26 	openvpn[76801]: cf_per = 0
Jan 29 20:58:26 	openvpn[76801]: cf_max = 0
Jan 29 20:58:26 	openvpn[76801]: push_ifconfig_ipv6_remote = ::
Jan 29 20:58:26 	openvpn[76801]: push_ifconfig_ipv6_local = ::/0
Jan 29 20:58:26 	openvpn[76801]: push_ifconfig_ipv6_defined = DISABLED
Jan 29 20:58:26 	openvpn[76801]: push_ifconfig_remote_netmask = 0.0.0.0
Jan 29 20:58:26 	openvpn[76801]: push_ifconfig_local = 0.0.0.0
Jan 29 20:58:26 	openvpn[76801]: push_ifconfig_defined = DISABLED
Jan 29 20:58:26 	openvpn[76801]: tmp_dir = '/tmp'
Jan 29 20:58:26 	openvpn[76801]: ccd_exclusive = DISABLED
Jan 29 20:58:26 	openvpn[76801]: client_config_dir = '/var/etc/openvpn-csc/1'

**Comment by user111222333** [comment]
There is a 'HARD REST' statements in logs....
There is something really wrong with this openvpn instance. 
There is nothing custom or out of ordinary, it's a vanilla OPNSense setup. 
It just doesn't work. 
Setting up by hand or using the wizard also did not make any difference. 
If some from a Dev team can try to set up the most basic openvpn instance on OPNsense 19.7.10, they should be able to see the same result we are getting here. 

Damien

**Comment by user111222333** [comment]
Increasing verbose level shows this: 

Jan 29 21:12:49 	openvpn[88704]: MANAGEMENT: Client disconnected
Jan 29 21:12:49 	openvpn[88704]: MANAGEMENT: CMD 'quit'
Jan 29 21:12:49 	openvpn[88704]: MANAGEMENT: CMD 'status 2'
Jan 29 21:12:48 	openvpn[88704]: MANAGEMENT: Client connected from /var/etc/openvpn/server1.sock
Jan 29 21:12:15 	openvpn[88704]: 10.0.90.107:58282 SIGUSR1[soft,tls-error] received, client-instance restarting
Jan 29 21:12:15 	openvpn[88704]: 10.0.90.107:58282 TLS Error: TLS handshake failed
Jan 29 21:12:15 	openvpn[88704]: 10.0.90.107:58282 TLS Error: TLS key negotiation failed to occur within 60 seconds (check your network connectivity)
Jan 29 21:11:47 	openvpn[88704]: MANAGEMENT: Client disconnected
Jan 29 21:11:47 	openvpn[88704]: MANAGEMENT: CMD 'quit'
Jan 29 21:11:47 	openvpn[88704]: MANAGEMENT: CMD 'status 2'
Jan 29 21:11:46 	openvpn[88704]: MANAGEMENT: Client connected from /var/etc/openvpn/server1.sock
Jan 29 21:11:44 	openvpn[88704]: 10.0.90.107:58282 UDPv4 WRITE [54] to [AF_INET]10.0.90.107:58282: P_CONTROL_HARD_RESET_SERVER_V2 kid=0 [ 3277609512 3435210917 864604340 4234456041 1850880880 1788553980 3022350159 1893129216 1630 845509120 0 ]
Jan 29 21:11:28 	openvpn[88704]: 10.0.90.107:58282 UDPv4 WRITE [54] to [AF_INET]10.0.90.107:58282: P_CONTROL_HARD_RESET_SERVER_V2 kid=0 [ 347884705 1569725851 1638275354 2193083082 1038272577 3740553228 1419249230 2093673728 1374 845509120 0 ]
Jan 29 21:11:20 	openvpn[88704]: 10.0.90.107:58282 UDPv4 WRITE [62] to [AF_INET]10.0.90.107:58282: P_ACK_V1 kid=0 [ 1423145466 948360812 3117878646 2242846258 3333216662 3562497691 3997957126 2569429248 1118 845509121 0 796657911 247663937 ]
Jan 29 21:11:20 	openvpn[88704]: 10.0.90.107:58282 UDPv4 READ [54] from [AF_INET]10.0.90.107:58282: P_CONTROL_HARD_RESET_CLIENT_V2 kid=0 [ 3376952937 3062836119 4260904155 2325310114 3725942554 2599974518 356847226 1306663680 862 845509120 0 ]
Jan 29 21:11:20 	openvpn[88704]: 10.0.90.107:58282 UDPv4 WRITE [54] to [AF_INET]10.0.90.107:58282: P_CONTROL_HARD_RESET_SERVER_V2 kid=0 [ 1008164119 1461544259 1725590468 1746142905 3834697833 1743586595 4086439986 3121669120 862 845509120 0 ]
Jan 29 21:11:16 	openvpn[88704]: 10.0.90.107:58282 UDPv4 WRITE [66] to [AF_INET]10.0.90.107:58282: P_CONTROL_HARD_RESET_SERVER_V2 kid=0 [ 3737670150 3762986542 3408401682 2852687415 761254607 2795678376 254640112 51360512 606 845509121 0 796657911 247663937 0 ]
Jan 29 21:11:16 	openvpn[88704]: 10.0.90.107:58282 UDPv4 READ [54] from [AF_INET]10.0.90.107:58282: P_CONTROL_HARD_RESET_CLIENT_V2 kid=0 [ 3553933340 1218841738 3843125938 2536631442 715718593 1354818686 3656827673 3600312576 606 845509120 0 ]
Jan 29 21:11:14 	openvpn[88704]: 10.0.90.107:58282 UDPv4 WRITE [66] to [AF_INET]10.0.90.107:58282: P_CONTROL_HARD_RESET_SERVER_V2 kid=0 [ 3815638749 1200520652 191780884 365297702 3123015444 3087908036 834477148 4139341568 350 845509121 0 796657911 247663937 0 ]
Jan 29 21:11:14 	openvpn[88704]: 10.0.90.107:58282 TLS: Initial packet from [AF_INET]10.0.90.107:58282, sid=2f7c08f7 0ec30d41
Jan 29 21:11:14 	openvpn[88704]: 10.0.90.107:58282 UDPv4 READ [54] from [AF_INET]10.0.90.107:58282: P_CONTROL_HARD_RESET_CLIENT_V2 kid=0 [ 3282989873 2544698115 2066731456 3847621881 44579705 3744312432 1330872439 1488519424 350 845509120 0 ]
Jan 29 21:11:14 	openvpn[88704]: 10.0.90.107:58282 Expected Remote Options String (VER=V4): 'V4,dev-type tun,link-mtu 1550,tun-mtu 1500,proto UDPv4,comp-lzo,keydir 1,cipher AES-256-GCM,auth [null-digest],keysize 256,tls-auth,key-method 2,tls-client'

**Comment by mimugmail** [comment]
Would you mind telling us the latest working version? There was no update from OPNsense UI or the pkg itself vom 19.7.9 to 19.7.10.

**Comment by user111222333** [comment]
Hello mimugmail, 
Last time i deployed OPNsense instance with a openvpn service was quite a while ago... over 10 months ago. It since got replaced with a dedicated vpn server behind a security perimeter (Debian + openvpn) and.... what was that version of OPNsense, I can't recall. But when it was working, the GUI config of OPNsense for the openvpn service looked a bit different. Client export tab had more options for clients configs for Apple IOS, Android, and a couple of desktop openvpn clients. None of those are present in this latest version of OPNSense. Looks like it got deprecated.

**Comment by user111222333** [comment]
More we looked at the logs during the attempts of the remote clients to connect to openvpn service, more it looks like openvpn service unable to check if the connecting client present in the local user DB. 
While client connecting, the 'connected client's' tab shows a connection in progress, but instead of identifying the username (users cert used for connection attempt), it just says 'unidentified'. 
So... at this point, this doesn't seem to be a openvpn service issue, it seems to be a deeper problem with the OPNsense internal user/users certificate management? 
This is it... at this point there is so much time spent on trying to get this to work with no result. 
OPNsense works as a router, firewall works well... it passes packets the way it is intended to and openvpn just going to have to be another dedicated machine on a local network. 
I hope dev team will be able to fix it in upcoming updates... because outside of this ... problem, this is an awesome software! 

Damien.

**Comment by mimugmail** [comment]
The dev team can't fix anything because OpenVPN works fine on 19.7.10. My own management server got an update from 19.7.5 to 19.7.10 and all clients reconnected fine. 

The best would be closing this issue here and open a thread in the forums with a couple of screenshots (server config, firewall rules, tcpdump, authentication) and we switch to there.

**Comment by mixmint** [comment]
In my case, it was enough to exchange the certificates for valid ones. It turned out that the previous ones I worked have expired. Unfortunately, this also required the replacement of all client keys. So pay attention to the validity of your certificates.

**Comment by user111222333** [comment]
Hello mixmint, 
This is a brand new installation of OPNsense from scratch, with all certs generated right after the install, so they are all valid in terms of time issued/expired variable.

Can you please comment if you have a working set up for openvpn using OPNsense as a base and what version you are using?

We just installed the latest OPNsense 20.1 from scratch, and the issue persist. 
Auto-generated config file(s) .ovpn, .p12, .key used for the clients config - do not result in successful connection to the server. 
Manually looking over of that's in generated .ovpn seems to be appropriate - and yet, no connection established.

Just to be consistent, the same client machine(s) where given a .ovpn configs generated by software called PFsense, and it went by the numbers with no issues. 

We also used the same client machines to test against a commercial instance of a OpenVPN server - no problems, client machine connects as expected. The network infrastructure remained the same, only variable that was changed is OPNsense hosting openvpn service and client config files where auto-generated by OPNsense hosted openvpn 'Client Export' tab in GUI. 

We have done what we could to get this to work, so far - no luck. ;(

Damien

**Comment by x-jokay** [comment]
I had a similar problem when using `Elliptic Curve` certs with `secp521r1` for a `peer-to-peer` setup of a customer. It did only work after changing the certs to use `secp384r1`, but you may use `RSA` as key type?

**Comment by user111222333** [comment]
We did set it up to use RSA and it was rather standard choices - RSA: AES-256 with SHA256. 
It looks like issue related not to crypto setup, it looks like server does not respond to client. 
Firewall rules all in place and look appropriate. We done this set up may time before with no issues... but  not in a most recent versions of OPNsense.

**Comment by x-jokay** [comment]
This is strange, I did multiple setups `peer-to-peer` and `remote access` using `19.7.9` without problems (beside `secp521r1`) four days ago.

**Comment by user111222333** [comment]
It is strange, and totally unexpected. We had a very, very good experience with OPNsense. I can confirm that we also did multiple deployments of OPNsense + openvpn service. Never had a problem until now.

**Comment by mixmint** [comment]
Hi Damien
Of course, here's my configuration looks like this:
OPNsense 19.7.10-amd64;
OpenVPN v2.4.8
On the server side:
Mode: Remote Access (SSL / TLS User Auth), Backend for authentication: local database, local group: none, UDP protocol, device: TUN, interface: in my case LAN (I also connect from the outside but I will not write how NAT is configured because the long history will come out), default server port
Peer Certificate Authority: here I had to generate a new one because it also expired ¯ \ _ (ツ) _ / ¯
Others are AES256CBC encryption, SHA1 auth, Cert Depth (Client + Server).
I use Redirect Gateway, Inter-client communication. I limit concurrent connections.
On the client side, the settings don't matter much to me.

I use two types of identical RSA certificates, one for the server and the other for clients, during generation I did not make any changes except those required in the Distinguished name, I used it so it was proposed. Maximum Lifetime. I noticed the variables in the Distinguished name and especially the Common Name to be identical for the server, client and most importantly Authorities certificate. I used to have a similar situation that I forgot about it on a certain mail server. I was looking all day for what I did wrong.

I have connected the new user certificate to the accounts in the local database. I have purchased packages with new keys and exchanged clients. Connections started this time without a problem (after the fifth attempt... I made some mistakes in Common Name  ¯ \ _ (ツ) _ / ¯ )

**Comment by user111222333** [comment]
Hello Mixmint, 
Thank you for your replay and details.  Our set up is very similar to your. 

So the issue has been mitigated... sort of. 
After not been able to get the clients to connect, we decided to take a drastic approach. 

We wiped the disk of the appliance and re-installed the OPNsense from scratch using the most up to date version 20.1-amd64.
Certificates where generated for server, for clients. opnevpn instance was configured... and this time, client machine connect as expected with out any issues.

Only difference this time, we did not just wiped the local openvpn service config and related certificates, we wiped out the whole base and started with a empty disk, installing everything from the scratch.

I would like to thank all who looked at this post(s) and helped us to figure this out.

OPNsense team - thank you for your development efforts, you guys are awesome! :)

Damien
